pyro/nn/auto_reg_nn.py

```diff
@@ -7,6 +7,75 @@
 from torch.nn import functional as F
 
 
+def sample_mask_indices(input_dim, hidden_dim, simple=False, conditional=False):
+    """
+    Samples the indices assigned to hidden units during the construction of MADE masks
+
+    :param input_dim: the dimensionality of the input variable
+    :type input_dim: int
+    :param hidden_dim: the dimensionality of the hidden layer
+    :type hidden_dim: int
+    :param simple: True to sample indices uniformly, false to space indices evenly and round up or down randomly
+    :type simple: bool
+    :param conditional: True if sampling indices for a conditional-MADE, false otherwise
+    :type conditional: bool
+    """
+    start_integer = 0 if conditional else 1
+
+    if simple:
+        # Simple procedure samples each index i.i.d.
+        return torch.randint(start_integer, input_dim, size=(hidden_dim,))
+    else:
+        # "Non-simple" procedure creates fractional indices evenly then rounds at random
+        mk = torch.linspace(start_integer, input_dim - 1, steps=hidden_dim)
+
+        # Randomly decide whether to round up or down indices
+        ints = mk.floor()
+        ints += torch.bernoulli(mk - ints)
+        return ints
+
+
+def create_mask(input_dim, observed_dim, hidden_dim, num_layers, permutation, output_dim_multiplier):
```

Comment:
in general can you please add tests that test the logic of `create_mask` separately?

Reply:
I think the logic of this is captured in the test for AutoRegressiveNN. If the mask is created incorrectly in any way then the Jacobian will not be triangular

Reply:
agreed that if there are "too many" connections then the jacobian will not be triangular but what if there are "too few"? how do we catch that? (might happen in some strange edge case, hidden dimension of 1 or 2 or what not)

Reply:
I think this could be handled by a separate test in `AutoRegressiveNNTests(TestCase)` that checks that all of the Jacobian above the diagaonal are nonzero

Reply:
I'm not sure how rigorous this would be though because if the weights in the network were small then these Jacobian elements could possibly underflow to 0

Reply:
well that's what it might be nice to test the masks more directly. because the jacobian being triangular means that things are 'correct' in the sense that one will get sensible log probabilities but it still might not be the expected behavior in terms of the connectivity patterns of the masks

Reply:
Okay, let me look into it

Reply:
Sorry for the delay - I've been thinking about this but haven't been able to decide on the best solution.

There are many different masks that are valid, it depends on how the indices are assigned to the hidden units. They could be sampled uniformly, evenly spaced and rounded, etc. It might not be best if the test was tied to a particular method.

Perhaps the test could start at the output and trace back through the masks to see which input units it's connected to? That way you could make sure that there aren't too few connections

Reply:
well, at a minimum i imagined checking that 
-- each output is connected to at least one input
-- each input is connected to at least one output
perhaps there are other reasonable checks on top of that?